include/pybind11/buffer_info.h

```diff
@@ -79,6 +72,18 @@ struct buffer_info {
     }
 
 private:
+    struct private_ctr_tag { explicit private_ctr_tag() = default; };
+
+    buffer_info(private_ctr_tag, void *ptr, ssize_t itemsize, const std::string &format, ssize_t ndim,
+                detail::any_container<ssize_t> &&shape_in, detail::any_container<ssize_t> &&strides_in)
+    : ptr(ptr), itemsize(itemsize), size(1), format(format), ndim(ndim),
+      shape(std::move(shape_in)), strides(std::move(strides_in)) {
+        if (ndim != (ssize_t) shape.size() || ndim != (ssize_t) strides.size())
+            pybind11_fail("buffer_info: ndim doesn't match shape and/or strides length");
+        for (size_t i = 0; i < (size_t) ndim; ++i)
+            size *= shape[i];
+    }
+
```

Comment:
Perhaps this could be redesigned a bit to only use this private constructor when you need the delegation, i.e. just in the one constructor at line 34 that wants to both move and call `->size`.  It makes the code a little easier to follow (one fewer delegation for most constructors), and invokes fewer moves.  It's also essentially a workaround for MSVC and old gcc, rather than a particularly desirable design.